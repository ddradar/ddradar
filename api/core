C:/Users/ebise/Documents/Projects/ddradar-azure/core