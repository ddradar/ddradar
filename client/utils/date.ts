export function unixTimeToString(unixTime: number) {
  const d = new Date(unixTime * 1000)
  return d.toLocaleString()
}
