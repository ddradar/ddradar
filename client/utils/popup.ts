import type { BuefyNamespace } from 'buefy/types'

const option = { position: 'is-top', hasIcon: true } as const

export function success(
  $buefy: Pick<BuefyNamespace, 'notification'>,
  message: string
) {
  $buefy.notification.open({ message, type: 'is-success', ...option })
}

export function warning(
  $buefy: Pick<BuefyNamespace, 'notification'>,
  message: string
) {
  $buefy.notification.open({ message, type: 'is-warning', ...option })
}

export function danger(
  $buefy: Pick<BuefyNamespace, 'notification'>,
  message: string
) {
  $buefy.notification.open({ message, type: 'is-danger', ...option })
}
