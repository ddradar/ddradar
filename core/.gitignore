# Generated from staticwebapp.database.schema.gql via graphql-codegen
src/graphql.ts
