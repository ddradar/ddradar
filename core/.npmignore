# TypeScript Files
tsconfig.json
**.ts
!dist/**.d.ts

# Test & Lint Files
.eslintrc.json
jest.config.**
/__tests__/**
