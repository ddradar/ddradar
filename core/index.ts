/* istanbul ignore file */

export * from './skill-attack'
export * from './typeUtils'
export * as Api from './api'
export * as Database from './db'
export * as Gate from './eagate-parser'
export * as Score from './score'
export * as Song from './song'
