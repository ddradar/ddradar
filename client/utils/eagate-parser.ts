import { getDanceLevel, UserScore } from '~/api/score'

type ChartScore = Omit<UserScore, 'userId' | 'userName'>

/**
 * Convert music data to { songId: Score[] } Record.
 * - https://p.eagate.573.jp/game/ddr/ddra20/p/playdata/music_data_single.html
 * - https://p.eagate.573.jp/game/ddr/ddra20/p/playdata/music_data_double.html
 */
export function musicDataToScoreList(
  sourceCode: string
): Record<string, ChartScore[]> {
  const doc = new DOMParser().parseFromString(sourceCode, 'text/html')
  const dataTable = doc.getElementById('data_tbl')
  if (!dataTable) throw new Error('invalid html')

  // Get PlayStyle from columns count
  const headerColumns =
    dataTable
      .getElementsByClassName('column')[0]
      ?.getElementsByClassName('rank')?.length ?? 0
  if (headerColumns !== 4 && headerColumns !== 5)
    throw new Error('invalid html')
  const playStyle = headerColumns === 5 ? 1 : 2

  const result: Record<string, ChartScore[]> = {}

  const songs = dataTable.getElementsByClassName('data')
  for (let i = 0; i < songs.length; i++) {
    const songRow = songs[i]

    // Get songName & songId from first column
    const songCol = songRow
      .getElementsByTagName('td')[0]
      .getElementsByClassName('music_info')[0]
    const songId = songCol
      .getAttribute('href')
      ?.replace(
        /^\/game\/ddr\/ddra20\/p\/playdata\/music_detail.html\?index=([01689bdiloqDIOPQ]{32})$/,
        '$1'
      )
    const songName = songCol.textContent?.trim()
    if (!songId || !songName) continue

    result[songId] = []

    const charts = songRow.getElementsByClassName('rank')
    for (let j = 0; j < charts.length; j++) {
      const chart = charts[j]

      // Get difficulty
      const difficulty = getDifficulty(chart.id.toLowerCase())

      // Get score
      const scoreText =
        chart.getElementsByClassName('data_score')[0].textContent ?? ''
      const score = parseInt(scoreText, 10)
      if (!Number.isInteger(score) || score < 0 || score > 1000000) continue

      // Get clearLamp
      const fcImageUrl = chart
        .getElementsByClassName('music_info')[0]
        .getElementsByTagName('img')[1]
        .src.toLowerCase()
      const rankImageUrl = chart
        .getElementsByClassName('music_info')[0]
        .getElementsByTagName('img')[0]
        .src.toLowerCase()
      const isFailed = /^.+\/rank_s_e\.png$/.test(rankImageUrl)
      const clearLamp =
        getClearLamp(fcImageUrl) ?? (isFailed ? 0 : score === 0 ? 1 : 2)

      // Get rank
      const rank = isFailed ? 'E' : getDanceLevel(score)

      result[songId].push({
        songId,
        songName,
        playStyle,
        difficulty,
        score,
        clearLamp,
        rank,
      })
    }
  }
  return result

  /** Get chart difficulty from element id */
  function getDifficulty(elementId: string) {
    if (elementId === 'beginner') return 0
    if (elementId === 'basic') return 1
    if (elementId === 'difficult') return 2
    if (elementId === 'expert') return 3
    if (elementId === 'challenge') return 4
    throw new Error('invalid html')
  }

  /** Get ClearLamp from full combo image URL */
  function getClearLamp(imageUrl: string) {
    // https://p.eagate.573.jp/game/ddr/ddra20/p/images/play_data/full_mar.png
    if (/^.+\/full_mar\.png$/.test(imageUrl)) return 7
    // https://p.eagate.573.jp/game/ddr/ddra20/p/images/play_data/full_perfect.png
    if (/^.+\/full_perfect\.png$/.test(imageUrl)) return 6
    // https://p.eagate.573.jp/game/ddr/ddra20/p/images/play_data/full_great.png
    if (/^.+\/full_great\.png$/.test(imageUrl)) return 5
    // https://p.eagate.573.jp/game/ddr/ddra20/p/images/play_data/full_good.png
    if (/^.+\/full_good\.png$/.test(imageUrl)) return 4
    // https://p.eagate.573.jp/game/ddr/ddra20/p/images/play_data/full_none.png
    return null
  }
}
