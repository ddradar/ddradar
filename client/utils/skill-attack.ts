import { getDanceLevel, UserScore } from '~/api/score'

type MusicScore = Omit<UserScore, 'songId' | 'userId' | 'userName' | 'level'>

/** Read shift-jis encoded text file asynchronously */
export function readTextAsync(file: File) {
  return new Promise<string>((resolve, reject) => {
    const reader = new FileReader()
    reader.onload = () => resolve(reader.result as string)
    reader.onerror = () => reject(reader.error)
    reader.readAsText(file, 'shift-jis')
  })
}

/**
 * Convert Skill Attack score list to ScoreList
 * - http://skillattack.com/sa4/data/dancer/{ddrCode}/score_{ddrCode}.txt
 */
export function scoreTexttoScoreList(text: string) {
  const rows = text.trim().split('\n')
  const result: Record<string, MusicScore[]> = {}
  rows.forEach(r => {
    /**
     * - [0]: skillAttackId
     * - [1]: 0:SP, 1:DP
     * - [2]: 0:BEGINNER, 1:BASIC, ..., 4:CHALLENGE
     * - [3]: DDR CODE
     * - [4]: UNIX Time
     * - [5]: score
     * - [6]: 0:Played(Failed,Assisted,Clear), 1:FC(include GreatFC), 2:PFC, 3:MFC
     * - [7]: Song Name (HTML escaped)
     * - [8]: blank
     */
    const cols = r.split('\t')
    if (cols.length < 8) return

    const skillAttackId = cols[0]
    const playStyle = (parseInt(cols[1], 10) + 1) as 1 | 2
    const difficulty = parseInt(cols[2], 10) as 0 | 1 | 2 | 3 | 4
    const score = parseInt(cols[5], 10)
    const clearLamp =
      cols[6] === '3' ? 7 : cols[6] === '2' ? 6 : cols[6] === '1' ? 4 : 2
    const songName = unescapeHTML(cols[7])!

    if (result[skillAttackId] === undefined) result[skillAttackId] = []

    const scores = result[skillAttackId]
    const oldScore = scores.find(
      s => s.playStyle === playStyle && s.difficulty === difficulty
    )
    if (oldScore === undefined) {
      scores.push({
        songName,
        playStyle,
        difficulty,
        clearLamp,
        score,
        rank: getDanceLevel(score),
      })
    } else {
      oldScore.clearLamp = clearLamp
      oldScore.score = score
      oldScore.rank = getDanceLevel(score)
    }
  })
  return result

  function unescapeHTML(escaped: string) {
    const doc = new DOMParser().parseFromString(escaped, 'text/html')
    return doc.documentElement.textContent
  }
}
