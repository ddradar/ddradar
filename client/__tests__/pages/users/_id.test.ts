import type { Api } from '@ddradar/core'
import { Database, Score } from '@ddradar/core'
import type { Context } from '@nuxt/types'
import {
  createLocalVue,
  mount,
  RouterLinkStub,
  shallowMount,
} from '@vue/test-utils'
import Buefy from 'buefy'
import { mocked } from 'ts-jest/utils'
import VueI18n from 'vue-i18n'

import {
  getClearStatus,
  getGrooveRadar,
  getScoreStatus,
  getUserInfo,
} from '~/api/user'
import UserPage from '~/pages/users/_id.vue'

jest.mock('~/api/user')

const localVue = createLocalVue()
localVue.use(Buefy)
localVue.use(VueI18n)

describe('/users/_id.vue', () => {
  const user: Api.UserInfo = {
    id: 'user_1',
    name: 'User 1',
    area: 13,
    code: 12345678,
  }
  const radar: Database.GrooveRadar = {
    stream: 120,
    voltage: 102,
    air: 160,
    freeze: 200,
    chaos: 103,
  }
  const clears = [...Array(19).keys()].map(i => ({
    level: i + 1,
    title: `${i + 1}`,
    statuses: [...Score.clearLampMap.keys(), -1 as const].map(clearLamp => ({
      clearLamp,
      count: (clearLamp + 2) * 10,
    })),
  }))
  const scores = [...Array(19).keys()].map(i => ({
    level: i + 1,
    title: `${i + 1}`,
    statuses: [...Score.danceLevelSet].map((rank, j) => ({
      rank,
      count: (j + 1) * 10,
    })),
  }))
  const i18n = new VueI18n({ locale: 'ja', silentFallbackWarn: true })
  const stubs = {
    NuxtLink: RouterLinkStub,
    ClearStatus: true,
    GrooveRadar: true,
    ScoreStatus: true,
  }
  const templateMocks = {
    $accessor: { user: null },
    $fetchState: { pending: false },
  }
  const data = () => ({
    user,
    radars: [radar, radar],
    clears: [clears, clears],
    scores: [scores, scores],
  })

  describe.each(['ja', 'en'])('{ locale: "%s" } snapshot test', locale => {
    const i18n = new VueI18n({ locale, silentFallbackWarn: true })

    test('renders correctly', () => {
      const mocks = { ...templateMocks }
      const wrapper = mount(UserPage, { localVue, i18n, stubs, mocks, data })
      expect(wrapper).toMatchSnapshot()
    })
    test('renders setting button if selfPage', () => {
      const mocks = { ...templateMocks, $accessor: { user: { id: user.id } } }
      const wrapper = mount(UserPage, { localVue, i18n, stubs, mocks, data })
      expect(wrapper).toMatchSnapshot()
    })
    test('renders empty if no data', () => {
      const mocks = { ...templateMocks }
      const wrapper = mount(UserPage, { localVue, i18n, stubs, mocks })
      expect(wrapper).toMatchSnapshot()
    })
    test('renders loading before fetch()', () => {
      const mocks = { ...templateMocks, $fetchState: { pending: true } }
      const wrapper = mount(UserPage, { localVue, i18n, stubs, mocks })
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('get areaName()', () => {
    test.each([
      ['ja', 0, '未指定'],
      ['en', 0, 'undefined'],
      ['ja', 47, '沖縄県'],
      ['en', 47, 'Okinawa'],
      ['ja', 51, 'アメリカ'],
      ['en', 51, 'USA'],
    ])('{ locale: "%s", area: %i } returns "%s"', (locale, area, expected) => {
      // Arrange
      const wrapper = shallowMount(UserPage, {
        localVue,
        i18n: new VueI18n({ locale, silentFallbackWarn: true }),
        stubs,
        mocks: templateMocks,
        data: () => ({ user: { ...user, area } }),
      })

      // Act - Assert
      // @ts-ignore
      expect(wrapper.vm.areaName).toBe(expected)
    })
    test.each(['ja', 'en'])(
      '{ locale: "%s" } returns "" if user is null',
      locale => {
        const i18n = new VueI18n({ locale, silentFallbackWarn: true })
        const mocks = { ...templateMocks }
        const wrapper = shallowMount(UserPage, { localVue, i18n, stubs, mocks })

        // Act - Assert
        // @ts-ignore
        expect(wrapper.vm.areaName).toBe('')
      }
    )
  })

  describe('get ddrCode()', () => {
    test.each([
      [0, ''],
      [10000000, '1000-0000'],
      [99999999, '9999-9999'],
    ])('{ code: %i } returns "%s"', (code, expected) => {
      // Arrange
      const wrapper = shallowMount(UserPage, {
        localVue,
        i18n,
        stubs,
        mocks: templateMocks,
        data: () => ({ user: { ...user, code } }),
      })

      // Act - Assert
      // @ts-ignore
      expect(wrapper.vm.ddrCode).toBe(expected)
    })
    test('returns "" if user is null', () => {
      const i18n = new VueI18n({ locale: 'ja', silentFallbackWarn: true })
      const wrapper = shallowMount(UserPage, {
        localVue,
        i18n,
        stubs,
        mocks: templateMocks,
      })

      // Act - Assert
      // @ts-ignore
      expect(wrapper.vm.areaName).toBe('')
    })
  })

  describe('get isSelfPage()', () => {
    test('returns false if no login', () => {
      // Arrange
      const wrapper = shallowMount(UserPage, {
        localVue,
        i18n,
        stubs,
        mocks: { ...templateMocks, $accessor: { user: null } },
        data: () => ({ user }),
      })

      // Act - Assert
      // @ts-ignore
      expect(wrapper.vm.isSelfPage).toBe(false)
    })
    test('returns false if id !== loginId', () => {
      // Arrange
      const id = 'user_2'
      const wrapper = shallowMount(UserPage, {
        localVue,
        i18n,
        stubs,
        mocks: { ...templateMocks, $accessor: { user: { id } } },
        data: () => ({ user }),
      })

      // Act - Assert
      // @ts-ignore
      expect(wrapper.vm.isSelfPage).toBe(false)
    })
    test('returns true if id === loginId', () => {
      // Arrange
      const wrapper = shallowMount(UserPage, {
        localVue,
        i18n,
        stubs,
        mocks: { ...templateMocks, $accessor: { user: { id: user.id } } },
        data: () => ({ user }),
      })

      // Act - Assert
      // @ts-ignore
      expect(wrapper.vm.isSelfPage).toBe(true)
    })
    test('returns false if user is null', () => {
      const i18n = new VueI18n({ locale: 'ja', silentFallbackWarn: true })
      const wrapper = shallowMount(UserPage, {
        localVue,
        i18n,
        stubs,
        mocks: templateMocks,
      })

      // Act - Assert
      // @ts-ignore
      expect(wrapper.vm.isSelfPage).toBe(false)
    })
  })

  describe('validate', () => {
    const mocks = { ...templateMocks }
    const wrapper = shallowMount(UserPage, { localVue, i18n, stubs, mocks })

    test.each(['', '@', 'あああ'])('({ id: "%s" }) returns false', id => {
      // Arrange
      const ctx = { params: { id } } as unknown as Context

      // Act - Assert
      expect(wrapper.vm.$options.validate!(ctx)).toBe(false)
    })
    test.each(['-_-', 'FOO', 'foo', '000'])(
      '({ id: "%s" }) returns true',
      id => {
        // Arrange
        const ctx = { params: { id } } as unknown as Context

        // Act - Assert
        expect(wrapper.vm.$options.validate!(ctx)).toBe(true)
      }
    )
  })

  describe('fetch()', () => {
    const $http = {}
    beforeEach(() => {
      mocked(getUserInfo).mockClear()
      mocked(getGrooveRadar).mockClear()
      mocked(getClearStatus).mockClear()
      mocked(getScoreStatus).mockClear()
    })

    test('/_id calls getUserInfo(this.$http, _id)', async () => {
      // Arrange
      const $route = { params: { id: 'foo' } }
      mocked(getUserInfo).mockResolvedValue(user)
      mocked(getGrooveRadar).mockResolvedValue([{ ...radar, playStyle: 1 }])
      mocked(getClearStatus).mockResolvedValue(
        [...Array(19 * (Score.clearLampMap.size + 1)).keys()].map(i => ({
          playStyle: 1,
          level: (i % 19) + 1,
          clearLamp: ((i % (Score.clearLampMap.size + 1)) - 1) as
            | Score.ClearLamp
            | -1,
          count: ((i % (Score.clearLampMap.size + 1)) + 1) * 10,
        }))
      )
      mocked(getScoreStatus).mockResolvedValue(
        [...Array(19 * (Score.danceLevelSet.size + 1)).keys()].map(i => ({
          playStyle: 1,
          level: (i % 19) + 1,
          rank:
            [...Score.danceLevelSet][i % (Score.danceLevelSet.size + 1)] ?? '-',
          count: ((i % (Score.danceLevelSet.size + 1)) + 1) * 10,
        }))
      )
      const mocks = { ...templateMocks, $http, $route }
      const wrapper = shallowMount(UserPage, { localVue, i18n, stubs, mocks })

      // Act
      await wrapper.vm.$options.fetch!.call(wrapper.vm, null!)

      // Assert
      expect(mocked(getUserInfo)).toBeCalledWith($http, $route.params.id)
      expect(wrapper.vm.$data.user).toBe(user)
    })
    test('sets user null if cause error', async () => {
      // Arrange
      const $route = { params: { id: 'foo' } }
      mocked(getUserInfo).mockRejectedValue('error')
      const mocks = { ...templateMocks, $http, $route }
      const wrapper = shallowMount(UserPage, { localVue, i18n, stubs, mocks })

      // Act
      await wrapper.vm.$options.fetch!.call(wrapper.vm, null!)

      // Assert
      expect(mocked(getUserInfo)).toBeCalledWith($http, $route.params.id)
      expect(wrapper.vm.$data.user).toBeNull()
    })
  })
})
