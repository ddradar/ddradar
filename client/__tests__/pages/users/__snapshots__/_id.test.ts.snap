// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/users/_id.vue snapshot test { locale: "en" } renders correctly 1`] = `
<section class="section">
  <h1 class="title">User 1</h1>
  <h2 class="subtitle">Tokyo / 1234-5678</h2>
  <!---->
</section>
`;

exports[`/users/_id.vue snapshot test { locale: "en" } renders empty 1`] = `<section class="section">User does not exist or is private</section>`;

exports[`/users/_id.vue snapshot test { locale: "en" } renders loading 1`] = `
<section class="section">
  <div class="b-skeleton is-animated">
    <div class="b-skeleton-item is-rounded"></div>
  </div>
  <div class="b-skeleton is-animated">
    <div class="b-skeleton-item is-rounded"></div>
  </div>
  <!---->
</section>
`;

exports[`/users/_id.vue snapshot test { locale: "en" } renders setting button if selfPage 1`] = `
<section class="section">
  <h1 class="title">User 1</h1>
  <h2 class="subtitle">Tokyo / 1234-5678</h2>
  <div class="buttons"><a type="button" class="button is-primary"><span class="icon is-small"><i class="mdi mdi-import"></i></span><span>
        Import Scores
      </span>
      <!---->
    </a> <a type="button" class="button is-info"><span class="icon is-small"><i class="mdi mdi-account-cog"></i></span><span>
        Settings
      </span>
      <!---->
    </a></div>
</section>
`;

exports[`/users/_id.vue snapshot test { locale: "ja" } renders correctly 1`] = `
<section class="section">
  <h1 class="title">User 1</h1>
  <h2 class="subtitle">東京都 / 1234-5678</h2>
  <!---->
</section>
`;

exports[`/users/_id.vue snapshot test { locale: "ja" } renders empty 1`] = `<section class="section">ユーザーが存在しないか、非公開に設定されています。</section>`;

exports[`/users/_id.vue snapshot test { locale: "ja" } renders loading 1`] = `
<section class="section">
  <div class="b-skeleton is-animated">
    <div class="b-skeleton-item is-rounded"></div>
  </div>
  <div class="b-skeleton is-animated">
    <div class="b-skeleton-item is-rounded"></div>
  </div>
  <!---->
</section>
`;

exports[`/users/_id.vue snapshot test { locale: "ja" } renders setting button if selfPage 1`] = `
<section class="section">
  <h1 class="title">User 1</h1>
  <h2 class="subtitle">東京都 / 1234-5678</h2>
  <div class="buttons"><a type="button" class="button is-primary"><span class="icon is-small"><i class="mdi mdi-import"></i></span><span>
        スコアのインポート
      </span>
      <!---->
    </a> <a type="button" class="button is-info"><span class="icon is-small"><i class="mdi mdi-account-cog"></i></span><span>
        設定
      </span>
      <!---->
    </a></div>
</section>
`;
